### ISSUE TITLE
Support "Inconclusive" as a test-result

### ISSUE THREAD

**Issue by petemounce** [fixie/fixie#57]
From #55, it would be useful to support a test-result of `Inconclusive`. I envision this being returned if the act of running the test threw an exception or returned an error, as opposed to the test itself.

For example
- if the test is trying to hit a web-service and that times out - that's a fail
- if the test is trying to hit a web-service and some configuration that the test relies on is missing - that's a fail
- if the test-runner thread is aborted for some (unforeseen!) reason, that's an inconclusive (it's not the test's fault, it's the runner)

My implementation of the `ICheckADependency` I mentioned in #55 has this as an abstract base class:

``` c#
    public abstract class DependencyChecker : ICheckADependency
    {
        public string Name { get; protected set; }

        public DependencyHealthCheckResponse Execute(DependencyHealthCheck request)
        {
            try
            {
                var isHealthy = IsHealthy();
                return isHealthy
                    ? DependencyHealthCheckResponse.Up(Name)
                    : Down;
            }
            catch (Exception ex)
            {
                return new DependencyHealthCheckResponse
                {
                    Name = Name,
                    Result = new Result<DependencyHealthCheckResults>(DependencyHealthCheckResults.ErrorPreventedConclusiveCheckResult, ex.Message),
                    Message = ex.Message,
                    Raw = ex.StackTrace
                };
            }
        }

        protected DependencyHealthCheckResponse Down { private get; set; }

        protected abstract bool IsHealthy();
    }
```

**Comment by Martin-Bohring** [comment]
I am creating a small specification framework on top of fixie (mostly a opinionated convention and some helper classes) and have run into the same situation (BDD, outside in testing etc.).

Is there a reason no one picked up this issue? I might want to give it a try.

**Comment by plioi** [comment]
@Martin-Bohring Sorry for my late reply.  Somehow I didn't get a notification.

There's no reason this hasn't been picked up yet.  It simply hasn't been a high enough priority for folks yet.

If you do, consider whether you want to start against master, which is stable and represents the 1.x releases, or dev, which is *quite* unstable work towards 2.x.

You'll probably be better off working against master aka 1.x, and let me worry about bringing it into 2.x once the dust settles there.

**Comment by Martin-Bohring** [comment]
**@plioi**  Sound like a plan.
I stilll need to come up with a good design idea.

- One idea could be to enhance  SkipBehavior and SkipResult with a typed SkipReason.
But that does not feel very SOLID to me (The Skip machinery now addressing 2 concerns)

- Another idea is to introduce an InconclusiveBehaviour and an InconclusiveResult.
That sounds much better to me, but might get a bit out of control in the long run as well
(More and more behaviours introduced over time)

Another issue is since Fixie is so nicely convention based, I would like not to introduce an attribute for inconclusive tests. Maybe having an empty test case implementation could be an indicator for inconclusive tests? Or an Inconclusive convention, that is registered within a Fixie Convention?

What do you think?

Do you have a topic branch naming scheme?

I don't know when I will get to (still struggling with a sample project for another project that is simple enough for everybody to understand, but still complex enough to demonstrate advantages)

But I plan to pick it up within the next 3 weeks.

**Comment by plioi** [comment]
You're right to avoid wanting to add an attribute for this. We let the end users define that sort of thing themselves, and instead offer up the more general concept for them to manipulate.

I don't really think it would be a bool by the time we're done, but at least start by picturing this as a new bool property living on Case.  A convention could then set it to true if they want the failure to be suppressed. That's at least a step in the right direction.

You'll find a CaseStatus enum as well, which would probably deserve a new Inconclusive value.  I recommend that before you get into coding, follow the existing flow of these CaseStatus values through the system, to see what all would be affected by a *new* one.

**Comment by plioi** [comment]
Also, it could help to get familiar with the "Expected Exception" part of the *sample* NUnit convention: https://github.com/fixie/fixie/blob/master/src/Fixie.Samples/NUnitStyle/CustomConvention.cs#L80

See how this manipulates a failing case so that it starts to pass.  And end user would want to accomplish Inconslusive state very similar to this, once the feature was supported.

**Comment by Martin-Bohring** [comment]
**@plioi**
I just forked the repository and found an already existing brach named **InconclusiveTests**
(unfortunately I don't know how to refer to branches in github flavoured markdown).

Some of the steps you have mentioned are already done on that branch. It is some 20 commits behind the master branch.

Does it make sense to continue on that branch, or is a fresh start from master the better option?

**Comment by plioi** [comment]
It's worth looking at it to see what my train of thought had been, but I imagine a lot has happened in master since then and it may be a dead end.  Feel free to use it as a guide, but I don't object to you starting a new branch against the latest master.

**Comment by Martin-Bohring** [comment]
**@plioi**
I started to work on the issue (I wish I could invest more time than I can right now)

**Are you interested in a WIP PR with a check list of what I plan to do?**
That would also force me to clear my thoughts :smile: 

Curremtly I am in a very early stage. I extendend the _CaseStatus_ enumeration and propagated it through the *Result classes (As you already have done on your branch)

I recognized there are no dedicated tests for the *Result classes.
Is this by intention? I guess those are tested with the other tests anyway.

But unitl the loop is closed completely, it might still be helpful to have some tests in this area.

What do you think?

**Comment by plioi** [comment]
Please feel free to open a WIP PR, as long as you remind me in the description that it's a work in progress.

I have a few heavy duty tasks on my plate in the short term, so I might not have much feedback early on.

The main way that the *Result types get tested is in the tests of the various *Listener classes themselves.  The more experimental 'dev' branch is much better about that, though.

**Comment by wojciech-kotlarski-insurwave** [comment]
Hello @plioi , I am looking to upgrade LightBDD.Fixie integration to support fixie 3.x  but it I have realised that I no longer can mark test cases ignored/skipped if they have been executed.

I have went through the https://github.com/fixie/fixie/wiki/Customizing-the-Test-Project-Lifecycle#recipe-skipping-tests documentation but both cases describes the scenario where actual test was never executed.

In case of LightBDD integration I need to be able to begin the execution of the test method because:
* by running the test, the LightBDD can record the test execution and include the result in the report,
* it is possible that test owners have implemented a part of BDD scenario but indicated that some of it's sub-steps are not implemented yet and the test execution should stop with inconclusive result at that stage.

I was able to implement such behaviour with Fixie v2 but not any more with Fixie v3.

To give more context, this is the typical LightBDD test structure for Fixie ([wiki link](https://github.com/LightBDD/LightBDD/wiki/Test-Framework-Integrations#lightbddfixie2)):
```c#
public partial class Basket_feature : FeatureFixture
{
    [Scenario]
    public void No_product_in_stock()
    {
        Runner.RunScenario(
            Given_product_is_out_of_stock,
            When_customer_adds_it_to_the_basket,
            Then_the_product_addition_should_be_unsuccessful,
            Then_the_basket_should_not_contain_the_product);
    }
}
```
The `Runner.RunScenario()` needs to be executed to record the scenario or control the inner steps execution, where these inner steps may request the scenario to be ignored.

The integration problem lays in the [Test.RunCore()](https://github.com/fixie/fixie/blob/main/src/Fixie/Test.cs#L170) method that has paths to `Pass` or `Fail` the test run, but there is no flow allowing to `Skip` test that is already executing.

If fixie would allow somehow to skip such test (via some internal control mechanism, dedicated SkipExecption or ability to override the test status after it was executed), then I would be able to migrate LightBDD.Fixie version.

I tried to use a workaround and call `test.Skip()` after the `test.Run()`, but it effected with 2 test runs being reported on Solution Test Explorer.

**Comment by Suremaker** [comment]
Hello @plioi , I am looking to upgrade LightBDD.Fixie integration to support fixie 3.x  but it I have realised that I no longer can mark test cases ignored/skipped if they have been executed.

I have went through the https://github.com/fixie/fixie/wiki/Customizing-the-Test-Project-Lifecycle#recipe-skipping-tests documentation but both cases describes the scenario where actual test was never executed.

In case of LightBDD integration I need to be able to begin the execution of the test method because:
* by running the test, the LightBDD can record the test execution and include the result in the report,
* it is possible that test owners have implemented a part of BDD scenario but indicated that some of it's sub-steps are not implemented yet and the test execution should stop with inconclusive result at that stage.

I was able to implement such behaviour with Fixie v2 but not any more with Fixie v3.

To give more context, this is the typical LightBDD test structure for Fixie ([wiki link](https://github.com/LightBDD/LightBDD/wiki/Test-Framework-Integrations#lightbddfixie2)):
```c#
public partial class Basket_feature : FeatureFixture
{
    [Scenario]
    public void No_product_in_stock()
    {
        Runner.RunScenario(
            Given_product_is_out_of_stock,
            When_customer_adds_it_to_the_basket,
            Then_the_product_addition_should_be_unsuccessful,
            Then_the_basket_should_not_contain_the_product);
    }
}
```
The `Runner.RunScenario()` needs to be executed to record the scenario or control the inner steps execution, where these inner steps may request the scenario to be ignored.

The integration problem lays in the [Test.RunCore()](https://github.com/fixie/fixie/blob/main/src/Fixie/Test.cs#L170) method that has paths to `Pass` or `Fail` the test run, but there is no flow allowing to `Skip` test that is already executing.

If fixie would allow somehow to skip such test (via some internal control mechanism, dedicated SkipExecption or ability to override the test status after it was executed), then I would be able to migrate LightBDD.Fixie version.

I tried to use a workaround and call `test.Skip()` after the `test.Run()`, but it effected with 2 test runs being reported on Solution Test Explorer.

**Comment by plioi** [comment]
Hi @Suremaker,

I do have a suggestion for something to try, but I'll start by giving some context on the full story. I think that will better explain the API and give you some sample code to start with for your solution.

In practice, Fixie v2's test class lifecycle hooks were fairly limited even for very commonly requested scenarios. It was especially painful for anyone trying to do interesting things with parameterized tests, where perfectly reasonable requests from users were *impossible*. This led to a pretty dramatic breaking change with v3 as you've experienced. Although I regret introducing the pain of such a breaking change, I regret v1 and v2's approach even more :). v1 and v2 were trying *too hard* to get in the way and provide tiny hooks to consumers, where v3 is essentially trying to get out of the way and allow consumers to write a block of plain idiomatic C# with full control.

As you're experiencing, that did come with a trade-off. The primary API calls (`await test.Run()`) *are* quite opinionated. To keep *most* consumers from having to do a lot of "bookkeeping" while providing a simple interface, we needed to better infer what was happening during execution, and that led to a more firm stance on what a test *is* and what the primary "boundary" of execution is: namely that a test is an invocation of a method, and the test is *absolutely 'done' (and safely reportable) once the method invocation is done*.

That's a real pain for folks who wanted to recharacterize what has happened after the fact.

I'll make an analogy with a similar feature from NUnit, and then I'll get to what you can try as a workaround since I think you'll have enough context and sample code by that point to give it a try.

A similar situation is NUnit-style `[ExpectedException(...)]` where the test method would be allowed to complete, and then the test framework would do more work to determine whether it *really* passed or *really* failed. Over time this approach was no longer recommended for NUnit users, since it was too easy to get a test that satisfied the attribute, even when the system under test was really throwing from a different specific line than the developer was intending to nail down. That led to things like *assertions* for a more narrow confirmation that a specific thing throws, rather than the whole test.

Basically, the conclusion was that it's better for the test framework to own and define what the boundary of execution is, and leave the user free to make their point *within* the method some other way. One can imagine a similar argument against quietly recharacterizing a failure as a skip, favoring instead rewriting the test to make the same *point* without relying on a specialized *throw* to do so. But to be clear, I recognize it is highly debatable and subjective rather than a silly statement of Right vs Wrong.

I bring up ExpectedException for a few reasons. First, it should start to sound familiar to your situation, in that we have the tension between wanting a test method's execution to be 1-1 with a test result, at least most of the time, and straying from that tends to thwart expectations. Second, it is a scenario I had in mind throughout development of v3, and having that in mind helped me to keep my "NUnit imitation" scenario working. That second point led to my inclusion of an "escape hatch" for folks who really want more control. You've found part of that API in the `test.Skip()` method, but understandably it's not super clear what to do to get that working the way you intended without double-reporting.

With the similar ExpectedException feature in mind, see https://github.com/fixie/fixie.integration/tree/main/src/NUnitStyle.Tests and `NUnitExecution ` in particular. It's... a lot. We call the raw event-reporting methods `test.Pass(...)`, `test.Fail(...)`, and `test.Skip(...)` explicitly, but note that we do *not* call the opinionated test-runing method `test.Run(...)`. `test.Run(...)` operates under the v3 stance that "running a test" means "executing the test fully and noting strictly whether or not it threw, so that reporting is automatic". Since that wouldn't be OK for ExpectedException, the NUnit behavior is met by instead call the test *method* directly and leaving the consumer `IExecution` to make sense of any exceptions/etc. The IExecution naturally has to do a lot more exception handling and control flow, as it's taking on more of Fixie's default job, but at least it has sufficient control. `await test.Method.Call(...)` is what you're looking for; it performs the same async-or-not handling so that you don't have to worry about that, but does not perform the "test" style exception catching/reporting, meaning that's left for you to do.

So, in summary:

1. Instead of calling `await test.Run(...)` to "run as a test" in the strict sense of the words, try the more fundamental `await test.Method.Call(...)` to "just call as a method" and then handle the possible exception yourself.
2. Be extra careful to emit a Pass, Fail, or Skip result according to your own desired logic. If you never call Fail, for instance, your customization would likely have a serious bug!
3. You might start with a copy of the NUnit example code, recognizing that "approximating a skip by checking for a specific expected exception" is quite similar in concept to NUnit's generalized ExpectedException concept. I think you'll find your solution can be a little simpler/shorter than the one you copy from, but not much shorter.

Please feel free to @ mention me into the LightBDD Issues if you have more questions, so the answers can be closer to where you are working.

**Comment by Suremaker** [comment]
Hello @plioi,

Thank you for the detailed response. I thought to keep this comment here so that other community members may leverage this knowledge as well.

The proposed solution worked for me and I am getting the proper behaviors for passed, skipped, and failed tests. Thank you!

It is also great to see it working nicely with Test Explorer and Resharper :)

I am happy with the solution and I am going to ship Fixie 3.x integration with LightBDD 3.7.0 (https://github.com/LightBDD/LightBDD/issues/292).

---

The minor note that I would like to make here is that there is no information on currently executing tests on Test Explorer or Resharper and they only pop up when the run is finished. I presume it may be related to the [`recorder.Start()`](https://github.com/fixie/fixie/blob/main/src/Fixie/Test.cs#L175) call that is made in `Test.Run()` method and which does not seem to be exposed publicly. 

It can be easily recreated if you make one of the [NUnitStyle Tests](https://github.com/fixie/fixie.integration/blob/main/src/NUnitStyle.Tests/CalculatorTests.cs#L31) take more time to execute by adding `Thread.Sleep(TimeSpan.FromSeconds(3));` for example.

If you add the same delay to [CustomConventions tests](https://github.com/fixie/fixie.integration/blob/main/src/CustomConvention.Tests/CalculatorTests.cs#L23) that does use `Run()` method under the hood, the test start is notified.

Exposing ability to notify test start will do the trick I presume.
